timeout in sentry code

```
rom https://cloud-images.ubuntu.com/daily/: rootfs: 99% (23.08MB/s)', ' - copying image for juju/xenial/amd64 from https://cloud-images.ubuntu.com/daily/: rootfs: 100% (23.18MB/s)', ' - copying image for juju/xenial/amd64 from https://cloud-images.ubuntu.com/daily/: rootfs: 100% (23.18MB/s)', '                                                                                                             ', ' - unable to get LXD image for juju/xenial/amd64: Hash mismatch for https://cloud-images.ubuntu.com/daily//server/xenial/20170620/xenial-server-cloudimg-amd64.squashfs: 1ca106f30241e7d4aba003923292116c229e8e811800cfeaa4a771e9e932f894 != 23271027b75765a7d46d44a7c8f9e7cf81526165856526b75c6007ea78346886', 'ERROR failed to bootstrap model: cannot start bootstrap instance: unable to get LXD image for juju/xenial/amd64: Hash mismatch for https://cloud-images.ubuntu.com/daily//server/xenial/20170620/xenial-server-cloudimg-amd64.squashfs: 1ca106f30241e7d4aba003923292116c229e8e811800cfeaa4a771e9e932f894 != 23271027b75765a7d46d44a7c8f9e7cf81526165856526b75c6007ea78346886']
2017-06-21 12:57:50,668 [DEBUG] conjure-up/hadoop-spark - events.py:50 - Setting Error at conjureup/events.py:133
2017-06-21 12:57:50,794 [ERROR] conjure-up/hadoop-spark - events.py:152 - Unhandled exception
Traceback (most recent call last):
  File "/snap/conjure-up/471/lib/python3.6/site-packages/conjureup/controllers/bootstrap/common.py", line 39, in do_bootstrap
    app.current_cloud_type))
Exception: Unable to bootstrap (cloud type: localhost)
2017-06-21 12:57:50,804 [DEBUG] conjure-up/hadoop-spark - __init__.py:32 - Showing dialog for exception: Unable to bootstrap (cloud type: localhost)
2017-06-21 12:57:54,810 [ERROR] conjure-up/hadoop-spark - sentry.errors: base.py:682 - Sentry responded with an error: <urlopen error timed out> (url: https://sentry.io/api/180147/store/)
Traceback (most recent call last):
  File "/snap/conjure-up/471/usr/lib/python3.6/urllib/request.py", line 1318, in do_open
    encode_chunked=req.has_header('Transfer-encoding'))
  File "/snap/conjure-up/471/usr/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/snap/conjure-up/471/usr/lib/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/snap/conjure-up/471/usr/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/snap/conjure-up/471/usr/lib/python3.6/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/snap/conjure-up/471/usr/lib/python3.6/http/client.py", line 964, in send
    self.connect()
  File "/snap/conjure-up/471/lib/python3.6/site-packages/raven/utils/http.py", line 31, in connect
    timeout=self.timeout,
  File "/snap/conjure-up/471/usr/lib/python3.6/socket.py", line 722, in create_connection
    raise err
  File "/snap/conjure-up/471/usr/lib/python3.6/socket.py", line 713, in create_connection
    sock.connect(sa)
socket.timeout: timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/snap/conjure-up/471/lib/python3.6/site-packages/raven/transport/threaded.py", line 172, in send_sync
    super(ThreadedHTTPTransport, self).send(url, data, headers)
  File "/snap/conjure-up/471/lib/python3.6/site-packages/raven/transport/http.py", line 43, in send
    ca_certs=self.ca_certs,
  File "/snap/conjure-up/471/lib/python3.6/site-packages/raven/utils/http.py", line 66, in urlopen
    return opener.open(url, data, timeout)
  File "/snap/conjure-up/471/usr/lib/python3.6/urllib/request.py", line 526, in open
    response = self._open(req, data)
  File "/snap/conjure-up/471/usr/lib/python3.6/urllib/request.py", line 544, in _open
    '_open', req)
  File "/snap/conjure-up/471/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/snap/conjure-up/471/lib/python3.6/site-packages/raven/utils/http.py", line 46, in https_open
    return self.do_open(ValidHTTPSConnection, req)
  File "/snap/conjure-up/471/usr/lib/python3.6/urllib/request.py", line 1320, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error timed out>
2017-06-21 12:57:54,815 [ERROR] conjure-up/hadoop-spark - sentry.errors.uncaught: base.py:704 - ['Exception: Unable to bootstrap (cloud type: localhost)', '  File "conjureup/controllers/bootstrap/common.py", line 39, in do_bootstrap']
2017-06-21 12:58:05,805 [DEBUG] conjure-up/hadoop-spark - events.py:50 - Setting Shutdown at conjureup/events.py:129
2017-06-21 12:58:05,807 [DEBUG] conjure-up/hadoop-spark - events.py:50 - Received Shutdown at conjureup/events.py:169
2017-06-21 12:58:05,807 [INFO] conjure-up/hadoop-spark - events.py:173 - Shutting down
```

--- a/conjureup/app.py
+++ b/conjureup/app.py
@@ -368,6 +368,7 @@ def main():
     app.sentry = raven.Client(
         dsn=SENTRY_DSN,
         release=VERSION,
+        transport=raven.transport.requests.RequestsHTTPTransport,
         processors=(
             'conjureup.utils.SanitizeDataProcessor',
         )

--- a/conjureup/controllers/bootstrap/common.py
+++ b/conjureup/controllers/bootstrap/common.py
@@ -34,7 +34,7 @@ class BaseBootstrapController:
             err_log = log_file.read_text('utf8').splitlines()
             app.log.error("Error bootstrapping controller: "
                           "{}".format(err_log))
-            app.sentry.context.merge({'extra': {'err_log': err_log}})
+            app.sentry.context.merge({'extra': {'err_log': err_log[-400:]}})
             raise Exception('Unable to bootstrap (cloud type: {})'.format(
                 app.current_cloud_type))
 

--- a/conjureup/controllers/deploystatus/common.py
+++ b/conjureup/controllers/deploystatus/common.py
@@ -1,5 +1,3 @@
-import asyncio
-
 from conjureup import events, utils
 from conjureup.app_config import app
 from conjureup.models.step import StepModel
@@ -11,22 +9,10 @@ async def wait_for_applications(msg_cb):
     app.log.info(msg)
     msg_cb(msg)
 
-    # retry done check a few times to work around
-    # https://bugs.launchpad.net/juju-wait/+bug/1680963
-    for i in range(3):
-        try:
-            step = StepModel({},
-                             filename='00_deploy-done',
-                             name='Deployment Watcher')
-            await utils.run_step(step,
-                                 msg_cb)
-            break
-        except Exception as e:
-            if i < 2 and 'Applications did not start successfully' in str(e):
-                await asyncio.sleep(5)
-                app.log.debug('Retrying 00_deploy-done: {}'.format(i))
-                continue
-            raise
+    step = StepModel({},
+                     filename='00_deploy-done',
+                     name='Deployment Watcher')
+    await utils.run_step(step, msg_cb)
 
     events.ModelSettled.set()
     msg = 'Model settled.'

--- a/conjureup/events.py
+++ b/conjureup/events.py
@@ -146,22 +146,11 @@ def handle_exception(loop, context):
         return  # already reporting an error
     Error.set()
     exc = context['exception']
+    exc_info = (type(exc), exc, exc.__traceback__)
 
     if not (app.noreport or any(pred(exc) for pred in NOTRACK_EXCEPTIONS)):
         track_exception(str(exc))
-        try:
-            exc_info = (type(exc), exc, exc.__traceback__)
-            app.sentry.captureException(exc_info, tags={
-                'spell': app.config.get('spell'),
-                'cloud_type': app.current_cloud_type,
-                'region': app.current_region,
-                'jaas': app.is_jaas,
-                'headless': app.headless,
-                'juju_version': utils.juju_version(),
-                'lxd_version': utils.lxd_version(),
-            })
-        except Exception:
-            app.log.exception('Error reporting error')
+        utils.sentry_report(exc_info=exc_info)
 
     app.log.exception('Unhandled exception', exc_info=exc)
 

--- a/conjureup/juju.py
+++ b/conjureup/juju.py
@@ -2,6 +2,7 @@
 """
 import asyncio
 import json
+import logging
 import os
 from concurrent import futures
 from pathlib import Path
@@ -12,7 +13,7 @@ import yaml
 from bundleplacer.charmstore_api import CharmStoreID
 from juju.model import Model
 
-from conjureup import consts, events
+from conjureup import consts, events, utils
 from conjureup.app_config import app
 from conjureup.utils import is_linux, juju_path, run, spew
 
@@ -333,7 +334,7 @@ def get_regions(cloud):
     if not isinstance(result, dict):
         msg = 'Unexpected response from regions: {}'.format(result)
         app.log.error(msg)
-        app.sentry.captureMessage(msg)
+        utils.sentry_report(msg, level=logging.ERROR)
         result = {}
     return result
 

--- a/conjureup/utils.py
+++ b/conjureup/utils.py
@@ -2,6 +2,7 @@ import asyncio
 import codecs
 import errno
 import json
+import logging
 import os
 import pty
 import re
@@ -162,14 +163,32 @@ async def run_step(step, msg_cb, event_name=None):
                         msg_cb(line)
                     await asyncio.sleep(0.01)
 
+    out_log = Path(step_path + '.out').read_text()
+    err_log = Path(step_path + '.err').read_text()
+
     if proc.returncode != 0:
-        status = run(['juju', 'status', '--format=yaml'])
         app.sentry.context.merge({'extra': {
-            'out_log': Path(step_path + '.out').read_text(),
-            'err_log': Path(step_path + '.err').read_text(),
-            'status': (status.stdout or b'').decode('utf8'),
+            'out_log_tail': out_log[-400:],
+            'err_log_tail': err_log[-400:],
         }})
         raise Exception("Failure in step {}".format(step.filename))
+
+    # special case for 00_deploy-done to report masked
+    # charm hook failures that were retried automatically
+    if not app.noreport:
+        failed_apps = set()  # only report each charm once
+        for line in err_log.splitlines():
+            if 'hook failure, will retry' in line:
+                log_leader = line.split()[0]
+                unit_name = log_leader.split(':')[-1]
+                app_name = unit_name.split('/')[0]
+                failed_apps.add(app_name)
+        for app_name in failed_apps:
+            # report each individually so that Sentry will give us a
+            # breakdown of failures per-charm in addition to per-spell
+            sentry_report('Retried hook failure', tags={
+                'app_name': app_name,
+            })
 
     if event_name is not None:
         track_event(event_name, "Done", "")
@@ -178,6 +197,45 @@ async def run_step(step, msg_cb, event_name=None):
         "conjure-up.{}.{}.result".format(app.config['spell'],
                                          step.filename))
     return (result or b'').decode('utf8')
+
+
+def sentry_report(message=None, exc_info=None, tags=None, **kwargs):
+    if app.noreport:
+        return
+
+    async def _sentry_report(event_type, kwargs):
+        try:
+            default_tags = {
+                'spell': app.config.get('spell'),
+                'cloud_type': app.current_cloud_type,
+                'region': app.current_region,
+                'jaas': app.is_jaas,
+                'headless': app.headless,
+                'juju_version': juju_version(),
+                'lxd_version': lxd_version(),
+            }
+
+            if message is not None and exc_info is None:
+                event_type = 'raven.events.Message'
+                kwargs['message'] = message
+                if 'level' not in kwargs:
+                    kwargs['level'] = logging.WARNING
+            else:
+                event_type = 'raven.events.Exception'
+                if exc_info is None or exc_info is True:
+                    kwargs['exc_info'] = sys.exc_info()
+                else:
+                    kwargs['exc_info'] = exc_info
+                if 'level' not in kwargs:
+                    kwargs['level'] = logging.ERROR
+
+            kwargs['tags'] = dict(default_tags, **(tags or {}))
+
+            app.sentry.capture(event_type, **kwargs)
+        except Exception:
+            app.log.exception('Error reporting error')
+
+    app.loop.create_task(_sentry_report())
 
 
 def can_sudo(password=None):

--- a/requirements.txt
+++ b/requirements.txt
@@ -2,7 +2,7 @@ aiofiles==0.3.1
 bundle-placement==0.0.1
 Jinja2==2.9.6
 juju==0.6.0
-juju-wait==2.5.1
+juju-wait==2.6.0
 prettytable==0.7.2
 progressbar2==3.20.0
 PyYAML==3.12

